hm_fashion_recs/fe.py

```diff
@@ -0,0 +1,247 @@
+        item_age_volumes = []
+        for age in range(16, 100):
+            low = age - age_shifts[age]
+            high = age + age_shifts[age]
+            tmp = (
+                tr.query("@low <= age <= @high")
+                .groupby("item")
+                .size()
+                .reset_index(name="age_volume")
+            )
+            tmp["age_volume"] = tmp["age_volume"].rank(ascending=False)
+            tmp["age"] = age
+            item_age_volumes.append(tmp)
+        item_age_volumes = pd.concat(item_age_volumes)
+        df = df.merge(item_age_volumes, on=["item", "age"], how="left")
+
+    with tm.timeit("user category most frequent"):
+        for c in ["department_no_idx"]:
+            tmp = pd.read_pickle(user_features_path / f"user_ohe_agg_week{week}_{c}.pkl")
+            cols = [c for c in tmp.columns if c != "user"]
+            # tmp = tmp[['user'] + cols]
+            tmp[cols] = tmp[cols] / tmp[cols].mean()
+            tmp[f"{c}_most_freq_idx"] = np.argmax(tmp[cols].values, axis=1)
+            df = df.merge(tmp[["user", f"{c}_most_freq_idx"]])
+
+    with tm.timeit("ohe dot products"):
+        item_target_cols = [c for c in items.columns if c.endswith("_idx")]
+
+        items_with_ohe = pd.get_dummies(
+            items[["item"] + item_target_cols], columns=item_target_cols
+        )
+
+        cols = [c for c in items_with_ohe.columns if c != "item"]
+        items_with_ohe[cols] = items_with_ohe[cols] / items_with_ohe[cols].mean()
+
+        users_with_ohe = users[["user"]]
+        for c in item_target_cols:
+            tmp = pd.read_pickle(user_features_path / f"user_ohe_agg_week{week}_{c}.pkl")
+            assert tmp["user"].tolist() == users_with_ohe["user"].tolist()
+            # tmp = tmp[['user'] + [c for c in tmp.columns if c.endswith('_mean')]]
+            tmp = tmp.drop("user", axis=1)
+            users_with_ohe = pd.concat([users_with_ohe, tmp], axis=1)
+
+        assert items_with_ohe["item"].tolist() == items["item"].tolist()
+        assert users_with_ohe["user"].tolist() == users["user"].tolist()
+
+        users_items = df[["user", "item"]].drop_duplicates().reset_index(drop=True)
+        n_split = 10
+        n_chunk = (len(users_items) + n_split - 1) // n_split
+        ohe = []
+        for i in range(0, len(users_items), n_chunk):
+            users_items_small = users_items.iloc[i : i + n_chunk].reset_index(drop=True)
+            users_small = users_items_small["user"].values
+            items_small = users_items_small["item"].values
+
+            for item_col in item_target_cols:
+                i_cols = [c for c in items_with_ohe.columns if c.startswith(item_col)]
+                u_cols = [f"user_ohe_agg_{c}" for c in i_cols]
+                print(users_with_ohe.shape)
+
+                users_items_small[f"{item_col}_ohe_score"] = (
+                    items_with_ohe[i_cols].values[items_small]
+                    * users_with_ohe[u_cols].values[users_small]
```

Comment:
I suppose removing conversion to numpy won't change anything
```suggestion
            users_small = users_items_small["user"]
            items_small = users_items_small["item"]

            for item_col in item_target_cols:
                i_cols = [c for c in items_with_ohe.columns if c.startswith(item_col)]
                u_cols = [f"user_ohe_agg_{c}" for c in i_cols]
                print(users_with_ohe.shape)

                users_items_small[f"{item_col}_ohe_score"] = (
                    items_with_ohe[i_cols][items_small]
                    * users_with_ohe[u_cols][users_small]
```

Reply:
Need to check this one.

Reply:
created an issue to investigate this #343

Reply:
Simple removal of `.values` will not work, additionally you will need to replace `[]` with `.loc[]` on line 244 and 245, then there is a problem with column names during multiplication, that can be fixed with `.set_axis`, then there is a problem with indexes during multiplication.

So I guess we leave that as is for optimization stage.